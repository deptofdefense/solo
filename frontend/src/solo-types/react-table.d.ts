import {
  UseExpandedHooks,
  UseExpandedInstanceProps,
  UseExpandedOptions,
  UseExpandedRowProps,
  UseExpandedState,
  UseFiltersColumnOptions,
  UseFiltersColumnProps,
  UseFiltersInstanceProps,
  UseFiltersOptions,
  UseFiltersState,
  UseGlobalFiltersInstanceProps,
  UseGlobalFiltersOptions,
  UseGlobalFiltersState,
  UsePaginationInstanceProps,
  UsePaginationOptions,
  UsePaginationState,
  UseRowStateCellProps,
  UseRowStateInstanceProps,
  UseRowStateOptions,
  UseRowStateRowProps,
  UseRowStateState,
  UseSortByColumnOptions,
  UseSortByColumnProps,
  UseSortByHooks,
  UseSortByInstanceProps,
  UseSortByOptions,
  UseSortByState
} from "react-table";

declare module "react-table" {
  export interface TableOptions<D extends object>
    extends UseExpandedOptions<D>,
      UseFiltersOptions<D>,
      UseGlobalFiltersOptions<D>,
      UsePaginationOptions<D>,
      UseRowStateOptions<D>,
      UseSortByOptions<D>,
      Record<string, any> {}

  export interface Hooks<D extends object = {}>
    extends UseExpandedHooks<D>,
      UseSortByHooks<D> {}

  export interface TableInstance<D extends object = {}>
    extends UseColumnOrderInstanceProps<D>,
      UseExpandedInstanceProps<D>,
      UseFiltersInstanceProps<D>,
      UseGlobalFiltersInstanceProps<D>,
      UsePaginationInstanceProps<D>,
      UseRowStateInstanceProps<D>,
      UseSortByInstanceProps<D> {}

  export interface TableState<D extends object = {}>
    extends UseColumnOrderState<D>,
      UseExpandedState<D>,
      UseFiltersState<D>,
      UseGlobalFiltersState<D>,
      UsePaginationState<D>,
      UseRowStateState<D>,
      UseSortByState<D> {}

  export interface Column<D extends object = {}>
    extends UseFiltersColumnOptions<D>,
      UseSortByColumnOptions<D> {}

  export interface ColumnInstance<D extends object = {}>
    extends UseFiltersColumnProps<D>,
      UseSortByColumnProps<D> {}

  export interface Cell<D extends object = {}>
    extends UseRowStateCellProps<D> {}

  export interface Row<D extends object = {}>
    extends UseExpandedRowProps<D>,
      UseRowStateRowProps<D> {}
}
