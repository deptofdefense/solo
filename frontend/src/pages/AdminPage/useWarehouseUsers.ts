import { useState, useCallback, useEffect } from "react";
import {
  PaginatedApiResponse,
  LoadingStatus,
  createFakeUsers,
  WarehouseUser
} from "solo-types";
import { useAuthContext } from "context";

const useWarehouseUsers = () => {
  const { apiCall } = useAuthContext();
  const [loadingStatus, setLoadingStatus] = useState<LoadingStatus>({
    loading: false
  });
  const [pageCount, setPageCount] = useState<number>(0);
  const [users, setUsers] = useState<WarehouseUser[]>([]);

  const fetchWarehouseUsers = useCallback(async () => {
    setLoadingStatus({
      loading: true,
      error: false
    });
    try {
      const { results, count } = await apiCall<
        PaginatedApiResponse<WarehouseUser[]>
      >("/warehouse/users/", {
        method: "GET"
      });
      setUsers(results);
      setPageCount(Math.ceil(count / 25));
      setLoadingStatus({
        loading: false
      });
    } catch (e) {
      setUsers(createFakeUsers(25));
      setPageCount(9);
      setLoadingStatus({
        loading: false
      });
    }
    // setLoadingStatus({
    //   loading: false,
    //   error: true,
    //   message: e.message || "Something went wrong"
    // })
  }, [setLoadingStatus, setUsers, apiCall]);

  useEffect(() => {
    fetchWarehouseUsers();
    // eslint-disable-next-line
  }, []);

  const modifyWarehouseUser = useCallback(
    (userId: number, data: Partial<WarehouseUser>) => {
      setUsers(
        users.map(user =>
          user.userId === userId
            ? { ...user, ...data, hasModified: true }
            : user
        )
      );
    },
    [setUsers, users]
  );

  const getUserById = useCallback(
    (userId: number) =>
      users.find(user => user.userId === userId) as WarehouseUser,
    [users]
  );

  const updateUserPermissons = useCallback(
    async (userId: number) => {
      modifyWarehouseUser(userId, {
        loading: true,
        error: false
      });
      try {
        const { canD6T, canCOR } = getUserById(userId);
        await apiCall(`/warehouse/users/${userId}/`, {
          method: "PATCH",
          body: JSON.stringify({
            canD6T,
            canCOR
          })
        });
        modifyWarehouseUser(userId, {
          loading: false
        });
      } catch (e) {
        modifyWarehouseUser(userId, {
          loading: false,
          error: true,
          message: e.message || "Something went wrong"
        });
      }
    },
    [apiCall, getUserById, modifyWarehouseUser]
  );

  return {
    loadingStatus,
    modifyWarehouseUser,
    updateUserPermissons,
    pageCount,
    users
  };
};

export default useWarehouseUsers;
