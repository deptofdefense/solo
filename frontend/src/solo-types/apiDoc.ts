import { ApiDocument, Status } from "solo-types";
import * as faker from "faker";

export const defaultApiDoc: ApiDocument = {
  id: 1,
  statuses: [
    {
      id: 1,
      dic: {
        id: 1,
        code: "AE1",
        desc: "in transit"
      },
      status_date: "2020-03-01T21:47:13-05:00",
      key_and_transmit_date: null,
      esd: "2020-03-20",
      projected_qty: 2,
      received_qty: null,
      document: 1
    },
    {
      id: 2,
      dic: {
        id: 2,
        code: "AS1",
        desc: "received"
      },
      status_date: "2020-03-05T21:47:56-05:00",
      key_and_transmit_date: null,
      esd: null,
      projected_qty: 2,
      received_qty: null,
      document: 1
    }
  ],
  suppadd: {
    id: 1,
    subinventorys: [
      {
        id: 1,
        locators: [
          {
            id: 1,
            code: "M1234AA",
            desc: "",
            subinventorys: 1
          },
          {
            id: 2,
            code: "M1235A",
            desc: "",
            subinventorys: 1
          },
          {
            id: 3,
            code: "M1236",
            desc: "",
            subinventorys: 1
          }
        ],
        code: "MTM_STGE",
        desc: "",
        suppadd: 1
      },
      {
        id: 2,
        locators: [],
        code: "MTM_LAY",
        desc: "",
        suppadd: 1
      },
      {
        id: 3,
        locators: [],
        code: "MTM_PEB",
        desc: "",
        suppadd: 1
      }
    ],
    code: "YMTM",
    desc: "Motor Transportation"
  },
  part: {
    id: 2,
    nsn: "5430015061999",
    nomen: "TANK,LIQUID STORAGE",
    uom: "ea",
    price: 15,
    sac: 1,
    serial_control_flag: "n",
    lot_control_flag: "n",
    recoverability_code: "z",
    shelf_life_code: 0,
    controlled_inv_item_code: "u"
  },
  service_request: {
    id: 1,
    service_request: "12345678"
  },
  addresses: [
    {
      id: 4,
      address_type: {
        id: 4,
        type: "Bill-To",
        desc: null
      },
      name: "AAC-M30300",
      ric: "SMS",
      addy1: "addy1",
      addy2: "addy2",
      addy3: "bldg 24006 Montezuma",
      city: "Arlington",
      state: "VA",
      zip: "22202",
      country: "United States",
      document: [1]
    },
    {
      id: 3,
      address_type: {
        id: 3,
        type: "Requester",
        desc: null
      },
      name: "requestor",
      ric: "5E3",
      addy1: "addy1",
      addy2: "addy2",
      addy3: "addy3",
      city: "Arlington",
      state: "VA",
      zip: "22202",
      country: "United States",
      document: [1]
    },
    {
      id: 2,
      address_type: {
        id: 2,
        type: "Ship-To",
        desc: null
      },
      name: "ship to",
      ric: "SMS",
      addy1: "addy1",
      addy2: "addy2",
      addy3: "addy3",
      city: "Arlington",
      state: "VA",
      zip: "22202",
      country: "United States",
      document: [1]
    },
    {
      id: 1,
      address_type: {
        id: 1,
        type: "Holder",
        desc: null
      },
      name: "AAC-M30300",
      ric: "SMS",
      addy1: "addy1",
      addy2: "addy2",
      addy3: "addy3",
      city: "Arlington",
      state: "VA",
      zip: "22202",
      country: "United States",
      document: [1]
    }
  ],
  sdn: "M3030012341234"
};

const idxToStatus = (idx: number): Status => {
  const base: Omit<Status, "dic"> = {
    id: Math.floor(Math.random() * 20),
    status_date: faker.date.recent().toISOString(),
    key_and_transmit_date: null,
    esd: "2020-03-20",
    projected_qty: 20,
    received_qty: faker.random.number(20),
    document: 1
  };
  switch (idx) {
    case 0:
      return {
        ...base,
        dic: {
          id: 1,
          code: "AE1",
          desc: "in transit"
        }
      };
    case 1:
      return {
        ...base,
        dic: {
          id: 2,
          code: "AS1",
          desc: "shipped"
        }
      };
    case 2:
      return {
        ...base,
        dic: {
          id: 3,
          code: "AS2",
          desc: "in transit"
        }
      };
    case 3:
      return {
        ...base,
        dic: {
          id: 4,
          code: "D6T",
          desc: "received"
        }
      };
    default:
      return {
        ...base,
        dic: {
          id: 5,
          code: "COR",
          desc: "confirmed"
        }
      };
  }
};

const createFakeApiDocuments = (count: number = 20): ApiDocument[] => {
  return Array.from({ length: count }).map(() => ({
    ...defaultApiDoc,
    statuses: Array.from({
      length: faker.random.number({ min: 1, max: 5 })
    }).map((_, idx) => idxToStatus(idx)),
    service_request: {
      id: faker.random.number(50),
      service_request: faker.random.alphaNumeric(8).toUpperCase()
    },
    part: {
      ...defaultApiDoc.part,
      nomen: faker.commerce.productName()
    },
    suppadd: {
      ...defaultApiDoc.suppadd,
      desc: faker.random.arrayElement([
        "Motor Transportation",
        "Armory",
        "Communications",
        "Electronics Maintenence"
      ])
    },
    addresses: defaultApiDoc.addresses.map(addy => ({
      ...addy,
      name: faker.company.companyName()
    }))
  }));
};

export default createFakeApiDocuments;
