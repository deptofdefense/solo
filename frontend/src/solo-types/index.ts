import { SortingRule, Filters } from "react-table";
export {
  default as createFakeApiDocs,
  defaultApiDoc,
  defaultApiResponse
} from "./apiDoc";
export * from "./warehouseUsers";
export { default as createFakeDocs, defaultDoc } from "./doc";

export interface Part {
  id: number;
  nsn: string;
  nomen: string;
  uom: string;
  price?: number;
  sac?: number;
  serial_control_flag?: string;
  lot_control_flag?: string;
  recoverability_code?: string;
  shelf_life_code?: number;
  controlled_inv_item_code?: string;
}

export interface Dic {
  id: number;
  code: string;
  desc: string;
}

export interface Status {
  id: number;
  dic: Dic;
  status_date: string;
  key_and_transmit_date?: string | null;
  esd?: string | null;
  received_qty: number | null;
  projected_qty: number;
  document: number;
  user?: number | null;
}

export interface AddressType {
  id: number;
  type: string;
  desc?: string | null;
}

export interface Address {
  id: number;
  address_type: AddressType;
  name: string;
  ric: string;
  addy1: string;
  addy2: string;
  addy3: string;
  city: string;
  state: string;
  zip: string;
  country: string;
  document: number[];
}

export interface ServiceRequest {
  id: number;
  service_request: string;
}

export interface Locator {
  id: number;
  code: string;
  desc?: string;
  subinventorys?: number;
}

export interface Subinventory {
  id: number;
  locators: Locator[];
  code: string;
  desc?: string;
  suppadd?: number;
}

export type LocatorMap = Record<string, Locator[]>;

export interface SuppAdd {
  id: number;
  code: string;
  subinventorys: Subinventory[];
}

interface BaseDocument {
  id: number;
  sdn: string;
  part: Part;
  statuses: Status[];
}

export interface ApiDocument extends BaseDocument {
  service_request: ServiceRequest;
  addresses: Address[];
  suppadd: SuppAdd;
}

export interface Document extends BaseDocument {
  serviceRequest: ServiceRequest;
  shipper?: Address;
  receiver?: Address;
  loadingStatus: LoadingStatus;
  subinventorys: Omit<Subinventory, "locators">[];
  locatorsBySubinventory: LocatorMap;
  commodityName: string;
  mostRecentStatusIdx: number;
  enteredReceivedQty: number;
  enteredReceivedBy: string;
  enteredSubinventoryCode: string;
  enteredLocatorCode: string;
}

export interface LoadingStatus {
  loading: boolean;
  error?: boolean;
  message?: string;
}

export interface Query<T extends object> {
  sort: SortingRule<T>[];
  page: number;
  filters: Filters<T>;
}

export interface PaginatedApiResponse<T> {
  results: T;
  count: number;
  next: number;
  previous: number;
}

export interface WarehouseUser extends LoadingStatus {
  userId: number;
  username: string;
  canD6T: boolean;
  canCOR: boolean;
  aac: string;
  hasModified?: boolean;
}
