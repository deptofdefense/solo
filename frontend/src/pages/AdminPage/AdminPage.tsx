import React, { useMemo } from "react";
import { Title, Table } from "components";
import createColumns from "./tableColumns";
import useWarehouseUsers from "./useWarehouseUsers";
import { WarehouseUser } from "solo-types";

const AdminPage: React.FC = () => {
  const {
    //loadingStatus,
    modifyWarehouseUser,
    updateUserPermissons,
    // pageCount,
    users
  } = useWarehouseUsers();

  const tableColumns = useMemo(
    () => createColumns(modifyWarehouseUser, updateUserPermissons),
    [modifyWarehouseUser, updateUserPermissons]
  );

  return (
    <div className="tablet:margin-x-8 overflow-x-auto">
      <Title>User Administration</Title>
      <Table<WarehouseUser> columns={tableColumns} data={users} />
    </div>
  );
};

export default AdminPage;
