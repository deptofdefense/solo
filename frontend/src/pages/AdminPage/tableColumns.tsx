import React from "react";
import { Column } from "react-table";
import { LoadingIcon } from "components";
import { WarehouseUser } from "solo-types";
import { Checkbox, Button } from "solo-uswds";

type CreateColumns = (
  modifyWarehouseUser: (userId: number, data: Partial<WarehouseUser>) => void,
  updateUserPermissons: (userId: number) => void
) => Column<WarehouseUser>[];

const createColumns: CreateColumns = (
  modifyWarehouseUser,
  updateUserPermissons
) => [
  {
    Header: "Loading",
    id: "loading",
    disableSortBy: true,
    Cell: ({
      row: {
        original: { loading, error }
      }
    }) => <LoadingIcon loading={loading} error={error} />
  },
  {
    Header: "AAC",
    accessor: "aac"
  },
  {
    Header: "User",
    accessor: "username"
  },
  {
    Header: "D6T",
    id: "canD6T",
    disableSortBy: true,
    Cell: ({
      row: {
        original: { canD6T, userId }
      }
    }) => {
      return (
        <div className="padding-left-2">
          <Checkbox
            data-testid="has-d6t-checkbox"
            checked={canD6T}
            onChange={e => {
              modifyWarehouseUser(userId, {
                canD6T: e.target.checked
              });
            }}
          />
        </div>
      );
    }
  },
  {
    Header: "COR",
    id: "canCOR",
    disableSortBy: true,
    Cell: ({
      row: {
        original: { userId, canCOR }
      }
    }) => {
      return (
        <div className="padding-left-2">
          <Checkbox
            data-testid="has-cor-checkbox"
            checked={canCOR}
            onChange={e => {
              modifyWarehouseUser(userId, {
                canCOR: e.target.checked
              });
            }}
          />
        </div>
      );
    }
  },
  {
    Header: "Submit Permissions",
    id: "submit",
    disableSortBy: true,
    Cell: ({
      row: {
        original: { userId, hasModified }
      }
    }) => {
      return (
        <div className="padding-left-2">
          <Button
            disabled={!hasModified}
            onClick={() => {
              updateUserPermissons(userId);
            }}
          >
            Submit
          </Button>
        </div>
      );
    }
  }
];

export default createColumns;
