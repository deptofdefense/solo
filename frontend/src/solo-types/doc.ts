import { Document } from "solo-types";

export const defaultDoc: Document = {
  id: 1,
  statuses: [
    {
      id: 1,
      dic: {
        id: 1,
        code: "AS1",
        desc: ""
      },
      status_date: "2020-03-01T21:47:13-05:00",
      key_and_transmit_date: null,
      esd: "2020-03-20",
      projected_qty: 2,
      received_qty: null,
      document: 1
    },
    {
      id: 2,
      dic: {
        id: 2,
        code: "AB1",
        desc: ""
      },
      status_date: "2020-03-05T21:47:56-05:00",
      key_and_transmit_date: null,
      esd: null,
      projected_qty: 2,
      received_qty: null,
      document: 1
    }
  ],
  part: {
    id: 2,
    nsn: "5430015061999",
    nomen: "TANK,LIQUID STORAGE",
    uom: "ea",
    price: 15,
    sac: 1,
    serial_control_flag: "n",
    lot_control_flag: "n",
    recoverability_code: "z",
    shelf_life_code: 0,
    controlled_inv_item_code: "u"
  },
  serviceRequest: {
    id: 1,
    service_request: "12345678"
  },
  suppadd: {
    id: 5,
    code: "YMTM",
    desc: "Motor Transportation",
    subinventorys: []
  },
  shipper: {
    id: 4,
    address_type: {
      id: 4,
      type: "Bill-To",
      desc: null
    },
    name: "AAC-M30300",
    ric: "SMS",
    addy1: "addy1",
    addy2: "addy2",
    addy3: "bldg 24006 Montezuma",
    city: "Arlington",
    state: "VA",
    zip: "22202",
    country: "United States",
    document: [1]
  },
  receiver: {
    id: 3,
    address_type: {
      id: 3,
      type: "Requester",
      desc: null
    },
    name: "requestor",
    ric: "5E3",
    addy1: "addy1",
    addy2: "addy2",
    addy3: "addy3",
    city: "Arlington",
    state: "VA",
    zip: "22202",
    country: "United States",
    document: [1]
  },
  sdn: "M3030012341234",
  commodityName: "Armory",
  loadingStatus: {
    loading: false
  },
  mostRecentStatusIdx: 1,
  enteredReceivedBy: "",
  enteredReceivedQty: 2
};

const createFakeDocuments = (count: number = 20): Document[] => {
  return Array.from({ length: count }).map(() => ({
    ...defaultDoc,
    id: Math.floor(Math.random() * 100)
  }));
};

export default createFakeDocuments;
