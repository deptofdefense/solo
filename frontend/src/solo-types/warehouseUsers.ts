import { WarehouseUser, PaginatedApiResponse } from "solo-types";
import * as faker from "faker";

export const defaultWareHouseUser: WarehouseUser = {
  userId: 1,
  username: "1234123123",
  first_name: "testFirst",
  last_name: "testLast",
  aac: "testaac",
  canD6T: false,
  canCOR: false,
  loading: false
};

export const defaultUserApiResponse: PaginatedApiResponse<WarehouseUser[]> = {
  results: [defaultWareHouseUser],
  count: 10,
  next: 2,
  previous: 1
};

export const createFakeUsers = (count: number = 25): WarehouseUser[] => {
  return Array.from({ length: count }).map((_, idx) => ({
    ...defaultWareHouseUser,
    userId: idx + 1,
    username: faker.finance.account(10),
    first_name: faker.name.firstName(),
    last_name: faker.name.lastName(),
    aac: faker.random.arrayElement(["M30300", "M30305"]),
    canCOR: faker.random.boolean(),
    canD6T: faker.random.boolean()
  }));
};
