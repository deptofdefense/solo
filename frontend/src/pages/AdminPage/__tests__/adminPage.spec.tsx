import React from "react";
import AdminPage from "../AdminPage";
import { render, fireEvent, wait } from "test-utils";
import { defaultUserApiResponse } from "solo-types";
//import { fireEvent } from "@testing-library/react";
//import { queryByAttribute, queryByTitle } from "@testing-library/react";
//import { wait } from "@testing-library/react";

describe("AdminPage Component", () => {
  const fetchMock = jest.fn();
  const defaultUser = defaultUserApiResponse.results[0];

  beforeEach(() => {
    fetchMock.mockResolvedValue(defaultUserApiResponse);
  });

  afterEach(() => {
    fetchMock.mockReset();
  });

  it("matches snapshot", async () => {
    const { asFragment, queryByText } = render(<AdminPage />, {
      authContext: {
        apiCall: fetchMock
      }
    });
    await wait(() => {
      expect(fetchMock).toHaveBeenCalled();
      expect(queryByText(defaultUser.username)).toBeInTheDocument();
      expect(queryByText(defaultUser.aac)).toBeInTheDocument();
    });
    expect(asFragment()).toMatchSnapshot();
  });

  /*

      fetchUsers errors
          error state in document
      
      fetchUsers success
          table loaded with data
      
      sumitUsers errors
          ^table must be loaded with at least 1 row
          error state for table row
      
      submitUsers succeeds
          ^table must be loaded with at least 1 row
          success state for table row

  */

  it("fetches users and places them in a table", async () => {
   fetchMock.mockResolvedValue({
     ...defaultUserApiResponse
   });
   const { queryByText } = render(<AdminPage />, {
     authContext: {
       apiCall: fetchMock
     }
   });
   await wait(() => {
     expect(fetchMock).toHaveBeenCalled();
     expect(fetchMock.mock.calls[0][0]).toEqual("/warehouse/users/");
     expect(fetchMock.mock.calls[0][1]).toMatchObject({
       method: "GET"
     });
     expect(queryByText(defaultUser.username)).toBeInTheDocument()
   });
  });

  it("fails to fetch users due to network error", async () => {
    fetchMock.mockRejectedValue(new Error("network error"));
    render(<AdminPage />, {
      authContext: {
        apiCall: fetchMock
      }
    });
    await wait(() => {
      expect(fetchMock).toHaveBeenCalled();
      // add check here once api is integrated
    });
  });

  it("click on checkboxes and change permissions", async () => {
    fetchMock.mockResolvedValue({
      ...defaultUserApiResponse,
      results: [{
        ...defaultUser,
        canD6T: false,
        canCOR: false,
      }]
    });
    const { getByTestId, getByText, queryByText, container } = render(<AdminPage />, {
      authContext: {
        apiCall: fetchMock
      } 
    });

    await wait(() => {
      expect(fetchMock).toHaveBeenCalled()
      expect(queryByText(defaultUser.username)).toBeInTheDocument()
    })
    const submit = getByText("Submit")
    expect(submit).toBeDisabled();

    const corCheckbox = getByTestId("has-cor-checkbox");
    const d6tCheckbox = getByTestId("has-d6t-checkbox")
    fireEvent.change(corCheckbox, {
      target: { checked: true }
    }); 
    fireEvent.change(d6tCheckbox, {
      target: { checked: true }
    }); 
    await wait(() => {
      expect(corCheckbox).toBeChecked() 
      expect(d6tCheckbox).toBeChecked()
    });

    fetchMock.mockResolvedValue({});
    fireEvent.click(submit);
    await wait(() => {
      expect(fetchMock).toHaveBeenCalledTimes(2)
      expect(fetchMock.mock.calls[1][0]).toEqual(`/warehouse/users/${defaultUser.id}`)
      expect(fetchMock.mock.calls[1][1]).toMatchObject({
        method: "PATCH",
        body: JSON.stringify({
          canD6T: true,
          canCOR: true
        })
      })
      expect(container.querySelector("svg.fa-check")).toBeInTheDocument()
    });
  });


  // it("modifies a user permission and submits causing an api call to backend", async () => {
  //   // setup fetch to return appropriate response
  //   // fetchMoock.mockResolved/Rejected value for initial data fetch
  //   // e.g. fetchMock.mockResolvedValue(defaultUserWarehouseApiResponse);
  //   // e.g. fetchMock.mockRejectedValue(new Error("network error"));

  //   // render component
  //     // expect stuff to look right

  //   // grab checkboxes you want 
  //     // checkbox = getAllBySomething

  //   // fireEvents on checkboxes you want
  //     // fireEvent.click checkboxes

  //   // expect checkboxes to be checked
  //     // await wait ( .... expect(checkox).toBeChecked()) 
      

  //   //fetchMock.mockResolved/rejected value
  
  //   // click submit
  //     // fireEvent.click(getBySomething('submit'))
    

  //   // expect api call / error state / success state
  //     // awaait wait (() expect(fetchMock).toHAveBeenCalledTimes(2))
  //     // expect(fetchMock.mock.calls[1][0]).toEqual('/modify/user/endpoint')
  //     // expect(fetchMock.mock.calls[1][1]).toMatchObject({
  //     //  userId: 1,
  //     //  permissions: ['D6T', 'COR', or whatever the data should look like sent to api]
  //     // })
  //     // expect(checkIcon).toBeInthedocument()
  //     //  or
  //     // expect(errorIcon).tobeinthedocument()
  //     // or
  //     // expect(loadingIcon).tobeinthedocument()

  //   // 
  // });





});
