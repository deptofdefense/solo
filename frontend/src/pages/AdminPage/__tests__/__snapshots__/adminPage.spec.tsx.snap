// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AdminPage Component matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="tablet:margin-x-8 overflow-x-auto"
  >
    <h1
      class="usa-prose"
    >
      User Administration
    </h1>
    <form
      class="grid-row flex-row flex-justify-center flex-align-center"
    >
      <div
        class="grid-col-full tablet:grid-col-2 margin-x-2"
      >
        <select
          class="usa-select"
          placeholder="Field"
        >
          <option
            value="M30300"
          >
            M30300
          </option>
          <option
            value="M30305"
          >
            M30305
          </option>
        </select>
      </div>
      <div
        class="grid-col-10 tablet:grid-col-3"
      >
        <input
          class="usa-input noOutline"
          placeholder="Add New User"
          type="text"
          value=""
        />
      </div>
      <div
        class="grid-col-full margin-x-2"
      >
        <p />
      </div>
    </form>
    <table
      class="usa-table grid-col-12 usa-table--borderless"
      role="table"
    >
      <thead>
        <tr
          role="row"
        >
          <th
            colspan="1"
            role="columnheader"
          >
            <div
              class="display-flex flex-justify flex-align-center flex-no-wrap text-no-wrap"
            >
              Loading
            </div>
          </th>
          <th
            colspan="1"
            role="columnheader"
            style="cursor: pointer;"
            title="Toggle SortBy"
          >
            <div
              class="display-flex flex-justify flex-align-center flex-no-wrap text-no-wrap"
            >
              AAC
              <svg
                aria-hidden="true"
                class="svg-inline--fa fa-chevron-up fa-w-14 margin-left-05"
                data-icon="chevron-up"
                data-prefix="fas"
                focusable="false"
                role="img"
                style="visibility: hidden;"
                viewBox="0 0 448 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M240.971 130.524l194.343 194.343c9.373 9.373 9.373 24.569 0 33.941l-22.667 22.667c-9.357 9.357-24.522 9.375-33.901.04L224 227.495 69.255 381.516c-9.379 9.335-24.544 9.317-33.901-.04l-22.667-22.667c-9.373-9.373-9.373-24.569 0-33.941L207.03 130.525c9.372-9.373 24.568-9.373 33.941-.001z"
                  fill="currentColor"
                />
              </svg>
            </div>
          </th>
          <th
            colspan="1"
            role="columnheader"
            style="cursor: pointer;"
            title="Toggle SortBy"
          >
            <div
              class="display-flex flex-justify flex-align-center flex-no-wrap text-no-wrap"
            >
              First Name
              <svg
                aria-hidden="true"
                class="svg-inline--fa fa-chevron-up fa-w-14 margin-left-05"
                data-icon="chevron-up"
                data-prefix="fas"
                focusable="false"
                role="img"
                style="visibility: hidden;"
                viewBox="0 0 448 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M240.971 130.524l194.343 194.343c9.373 9.373 9.373 24.569 0 33.941l-22.667 22.667c-9.357 9.357-24.522 9.375-33.901.04L224 227.495 69.255 381.516c-9.379 9.335-24.544 9.317-33.901-.04l-22.667-22.667c-9.373-9.373-9.373-24.569 0-33.941L207.03 130.525c9.372-9.373 24.568-9.373 33.941-.001z"
                  fill="currentColor"
                />
              </svg>
            </div>
          </th>
          <th
            colspan="1"
            role="columnheader"
            style="cursor: pointer;"
            title="Toggle SortBy"
          >
            <div
              class="display-flex flex-justify flex-align-center flex-no-wrap text-no-wrap"
            >
              Last Name
              <svg
                aria-hidden="true"
                class="svg-inline--fa fa-chevron-up fa-w-14 margin-left-05"
                data-icon="chevron-up"
                data-prefix="fas"
                focusable="false"
                role="img"
                style="visibility: hidden;"
                viewBox="0 0 448 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M240.971 130.524l194.343 194.343c9.373 9.373 9.373 24.569 0 33.941l-22.667 22.667c-9.357 9.357-24.522 9.375-33.901.04L224 227.495 69.255 381.516c-9.379 9.335-24.544 9.317-33.901-.04l-22.667-22.667c-9.373-9.373-9.373-24.569 0-33.941L207.03 130.525c9.372-9.373 24.568-9.373 33.941-.001z"
                  fill="currentColor"
                />
              </svg>
            </div>
          </th>
          <th
            colspan="1"
            role="columnheader"
            style="cursor: pointer;"
            title="Toggle SortBy"
          >
            <div
              class="display-flex flex-justify flex-align-center flex-no-wrap text-no-wrap"
            >
              DODID
              <svg
                aria-hidden="true"
                class="svg-inline--fa fa-chevron-up fa-w-14 margin-left-05"
                data-icon="chevron-up"
                data-prefix="fas"
                focusable="false"
                role="img"
                style="visibility: hidden;"
                viewBox="0 0 448 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M240.971 130.524l194.343 194.343c9.373 9.373 9.373 24.569 0 33.941l-22.667 22.667c-9.357 9.357-24.522 9.375-33.901.04L224 227.495 69.255 381.516c-9.379 9.335-24.544 9.317-33.901-.04l-22.667-22.667c-9.373-9.373-9.373-24.569 0-33.941L207.03 130.525c9.372-9.373 24.568-9.373 33.941-.001z"
                  fill="currentColor"
                />
              </svg>
            </div>
          </th>
          <th
            colspan="1"
            role="columnheader"
          >
            <div
              class="display-flex flex-justify flex-align-center flex-no-wrap text-no-wrap"
            >
              D6T
            </div>
          </th>
          <th
            colspan="1"
            role="columnheader"
          >
            <div
              class="display-flex flex-justify flex-align-center flex-no-wrap text-no-wrap"
            >
              COR
            </div>
          </th>
          <th
            colspan="1"
            role="columnheader"
          >
            <div
              class="display-flex flex-justify flex-align-center flex-no-wrap text-no-wrap"
            >
              Submit Permissions
            </div>
          </th>
        </tr>
      </thead>
      <tbody
        role="rowgroup"
      >
        <tr
          role="row"
        >
          <td
            class="td"
            role="cell"
          >
            <span>
              <svg
                aria-hidden="true"
                class="svg-inline--fa fa-check fa-w-16 "
                data-icon="check"
                data-prefix="fas"
                focusable="false"
                role="img"
                viewBox="0 0 512 512"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M173.898 439.404l-166.4-166.4c-9.997-9.997-9.997-26.206 0-36.204l36.203-36.204c9.997-9.998 26.207-9.998 36.204 0L192 312.69 432.095 72.596c9.997-9.997 26.207-9.997 36.204 0l36.203 36.204c9.997 9.997 9.997 26.206 0 36.204l-294.4 294.401c-9.998 9.997-26.207 9.997-36.204-.001z"
                  fill="currentColor"
                />
              </svg>
            </span>
          </td>
          <td
            class="td"
            role="cell"
          >
            testaac
          </td>
          <td
            class="td"
            role="cell"
          >
            testFirst
          </td>
          <td
            class="td"
            role="cell"
          >
            testLast
          </td>
          <td
            class="td"
            role="cell"
          >
            1234123123
          </td>
          <td
            class="td"
            role="cell"
          >
            <div
              class="padding-left-2"
            >
              <input
                class="usa-checkbox__input inp"
                data-testid="has-d6t-checkbox"
                type="checkbox"
              />
            </div>
          </td>
          <td
            class="td"
            role="cell"
          >
            <div
              class="padding-left-2"
            >
              <input
                class="usa-checkbox__input inp"
                data-testid="has-cor-checkbox"
                type="checkbox"
              />
            </div>
          </td>
          <td
            class="td"
            role="cell"
          >
            <div
              class="padding-left-2"
            >
              <button
                class="usa-button"
                disabled=""
              >
                Submit
              </button>
            </div>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</DocumentFragment>
`;
